File: focus-core/src/commonMain/kotlin/org/bilibilifocus/core/service/DynamicFeedService.kt
package org.bilibilifocus.core.service

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.bilibilifocus.core.cookie.Cookie
import org.bilibilifocus.core.cookie.CookieProvider
import org.bilibilifocus.core.model.DynamicCard

class DynamicFeedService(
    private val cookieProvider: CookieProvider,
    private val httpClient: HttpClient,
    private val timezoneOffsetMinutes: Int = currentTimezoneOffsetMinutes(),
    private val userAgent: String = "Mozilla/5.0 (Linux; Android 10) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Mobile Safari/537.36",
) {
    data class FeedPage(
        val cards: List<DynamicCard>,
        val nextOffset: String?,
    )

    sealed class ServiceError(message: String) : Exception(message) {
        data object LoginRequired : ServiceError("需要登录或登录已失效")
        data object InvalidResponse : ServiceError("动态接口返回无效数据")
        data class Api(val code: Int, override val message: String) :
            ServiceError(message)
    }

    private val json = Json { ignoreUnknownKeys = true }

    suspend fun fetchFollowingFeed(): List<DynamicCard> =
        fetchFollowingFeedPage().cards

    suspend fun fetchFollowingFeedPage(offset: String? = null): FeedPage {
        val cookies = cookieProvider.loadCookies()
        if (cookies.isEmpty()) throw ServiceError.LoginRequired

        val url = buildEndpoint(offset)
        val headers = buildHeaders(cookies)
        val response = httpClient.get(url, headers)

        if (response.statusCode !in 200..299) {
            throw ServiceError.InvalidResponse
        }

        return decodeFeedPage(response.body)
    }

    private fun buildEndpoint(offset: String?): String {
        val sb = StringBuilder("https://api.bilibili.com/x/polymer/web-dynamic/v1/feed/all")
        sb.append("?type=all")
        sb.append("&timezone_offset=$timezoneOffsetMinutes")
        if (!offset.isNullOrEmpty()) {
            sb.append("&offset=$offset")
        }
        return sb.toString()
    }

    private fun buildHeaders(cookies: List<Cookie>): Map<String, String> {
        return mapOf(
            "Referer" to "https://t.bilibili.com/",
            "Accept" to "application/json, text/plain, */*",
            "User-Agent" to userAgent,
        ).let { cookieProvider.attachCookies(it, cookies) }
    }

    fun decodeFeedPage(rawJson: String): FeedPage {
        val payload = try {
            json.parseToJsonElement(rawJson) as? JsonObject
        } catch (_: Exception) {
            throw ServiceError.InvalidResponse
        } ?: throw ServiceError.InvalidResponse

        val code = payload.intValueAt("code") ?: -1
        val message = payload.stringValueAt("message") ?: "unknown"

        if (code == -101) throw ServiceError.LoginRequired
        if (code != 0) throw ServiceError.Api(code, message)

        val items = payload.arrayValueAt("data", "items")
        val cards = items?.mapNotNull { element ->
            val dict = element as? JsonObject ?: return@mapNotNull null
            makeCard(dict)
        } ?: emptyList()

        val nextOffset = payload.stringValueAt("data", "offset")
            ?.trim()
            ?.takeIf { it.isNotEmpty() }

        return FeedPage(cards, nextOffset)
    }

    private fun makeCard(item: JsonObject): DynamicCard? {
        val modules = item.dictionaryValueAt("modules") ?: JsonObject(emptyMap())
        val author = modules.dictionaryValueAt("module_author") ?: JsonObject(emptyMap())
        val dynamic = modules.dictionaryValueAt("module_dynamic") ?: JsonObject(emptyMap())
        val major = dynamic.dictionaryValueAt("major") ?: JsonObject(emptyMap())
        val basic = item.dictionaryValueAt("basic") ?: JsonObject(emptyMap())

        val id = item.stringValueAt("id_str")
            ?: basic.stringValueAt("comment_id_str")
            ?: basic.stringValueAt("rid_str")
            ?: basic.intValueAt("comment_id")?.toString()
            ?: return null

        val authorName = author.stringValueAt("name") ?: "Bilibili"
        val authorMid = author.intValueAt("mid")?.toLong() ?: 0L
        val authorAvatar = normalizedURL(author.stringValueAt("face"))
        val publishTime = author.stringValueAt("pub_time")
            ?: author.stringValueAt("pub_action")
            ?: author.stringValueAt("pub_time_label")
            ?: ""

        val majorType = major.stringValueAt("type") ?: ""

        val jumpCandidates = listOfNotNull(
            normalizedURL(basic.stringValueAt("jump_url")),
            normalizedURL(major.stringValueAt("archive", "jump_url")),
            normalizedURL(major.stringValueAt("article", "jump_url")),
            normalizedURL(major.stringValueAt("pgc", "jump_url")),
            normalizedURL(major.stringValueAt("courses", "jump_url")),
            normalizedURL(major.stringValueAt("music", "jump_url")),
            normalizedURL(major.stringValueAt("medialist", "jump_url")),
            normalizedURL(major.stringValueAt("live", "jump_url")),
            normalizedURL(major.stringValueAt("opus", "jump_url")),
            normalizedURL(major.stringValueAt("common", "jump_url")),
        ) + extractJumpURLs(major)

        val coverURLs = makeCoverURLs(major)
        val text = makeText(dynamic, major)
        val videoURL = jumpCandidates.firstOrNull { isVideoLikeURL(it) }
        val kind = makeKind(majorType, item.stringValueAt("type") ?: "", videoURL, coverURLs)
        val targetURL = videoURL
            ?: jumpCandidates.firstOrNull()
            ?: fallbackTargetURL(id, kind)

        return DynamicCard(
            id = id,
            kind = kind,
            author = DynamicCard.Author(name = authorName, mid = authorMid, avatarURL = authorAvatar),
            publishTime = publishTime,
            text = text,
            coverURLs = coverURLs,
            targetURL = targetURL,
            videoURL = videoURL,
        )
    }

    private fun makeText(dynamic: JsonObject, major: JsonObject): String {
        val candidates = listOfNotNull(
            dynamic.stringValueAt("desc", "text"),
            major.stringValueAt("opus", "summary", "text"),
            major.stringValueAt("archive", "title"),
            major.stringValueAt("article", "title"),
            major.stringValueAt("pgc", "title"),
            major.stringValueAt("courses", "title"),
            major.stringValueAt("music", "title"),
            major.stringValueAt("medialist", "title"),
            major.stringValueAt("live", "title"),
            major.stringValueAt("common", "title"),
        ) + extractTextCandidates(major)

        return candidates
            .mapNotNull { it?.trim()?.takeIf { t -> t.isNotEmpty() } }
            .firstOrNull() ?: ""
    }

    private fun makeCoverURLs(major: JsonObject): List<String> {
        val urls = mutableListOf<String>()

        normalizedURL(major.stringValueAt("archive", "cover"))?.let { urls.add(it) }
        major.arrayValueAt("article", "covers")?.forEach { cover ->
            (cover as? kotlinx.serialization.json.JsonPrimitive)?.content?.let { normalizedURL(it) }?.let { urls.add(it) }
        }
        normalizedURL(major.stringValueAt("pgc", "cover") ?: major.stringValueAt("pgc", "ep_cover"))
            ?.let { urls.add(it) }
        normalizedURL(major.stringValueAt("courses", "cover"))?.let { urls.add(it) }
        normalizedURL(major.stringValueAt("music", "cover"))?.let { urls.add(it) }
        normalizedURL(major.stringValueAt("medialist", "cover"))?.let { urls.add(it) }
        normalizedURL(
            major.stringValueAt("live", "cover") ?: major.stringValueAt("live", "room_cover")
        )?.let { urls.add(it) }

        major.arrayValueAt("draw", "items")?.forEach { item ->
            val dict = item as? JsonObject ?: return@forEach
            normalizedURL(dict.stringValueAt("src") ?: dict.stringValueAt("url"))?.let { urls.add(it) }
        }

        major.arrayValueAt("opus", "pics")?.forEach { picture ->
            val dict = picture as? JsonObject ?: return@forEach
            normalizedURL(dict.stringValueAt("url") ?: dict.stringValueAt("src"))?.let { urls.add(it) }
        }

        normalizedURL(major.stringValueAt("common", "cover"))?.let { urls.add(it) }
        urls.addAll(extractImageURLs(major))

        return urls.distinct()
    }

    private fun makeKind(
        majorType: String,
        itemType: String,
        videoURL: String?,
        coverURLs: List<String>,
    ): DynamicCard.Kind {
        if (videoURL != null ||
            itemType.contains("_AV") ||
            majorType in setOf(
                "MAJOR_TYPE_ARCHIVE", "MAJOR_TYPE_PGC",
                "MAJOR_TYPE_COURSES", "MAJOR_TYPE_MEDIALIST",
            )
        ) {
            return DynamicCard.Kind.VIDEO
        }

        return when (majorType) {
            "MAJOR_TYPE_DRAW", "MAJOR_TYPE_OPUS" ->
                if (coverURLs.isEmpty()) DynamicCard.Kind.TEXT else DynamicCard.Kind.IMAGE
            "MAJOR_TYPE_ARTICLE" -> DynamicCard.Kind.ARTICLE_LIKE
            else -> if (coverURLs.isEmpty()) DynamicCard.Kind.TEXT else DynamicCard.Kind.IMAGE
        }
    }

    private fun normalizedURL(rawValue: String?): String? {
        if (rawValue == null) return null
        var value = rawValue.trim()
        if (value.isEmpty()) return null

        value = when {
            value.startsWith("//") -> "https:$value"
            value.startsWith("/opus/") || value.startsWith("/video/") ||
                value.startsWith("/bangumi/play/") -> "https://www.bilibili.com$value"
            value.startsWith("/") -> "https://t.bilibili.com$value"
            else -> value
        }

        return org.bilibilifocus.core.routing.FocusNavigationPolicy.canonicalWebURL(value)
    }

    private fun isVideoLikeURL(url: String): Boolean {
        val path = url.lowercase().let { u ->
            u.substringAfter("bilibili.com", u).substringBefore("?")
        }
        return path.startsWith("/video/") || path.startsWith("/bangumi/play/")
    }

    private fun extractJumpURLs(major: JsonObject): List<String> {
        return extractURLs(major, urlKeyPatterns, maxDepth = 4) { isLikelyNavigationURL(it) }
    }

    private fun extractImageURLs(major: JsonObject): List<String> {
        return extractURLs(major, imageKeyPatterns, maxDepth = 4) { isLikelyImageURL(it) }
    }

    private fun extractTextCandidates(major: JsonObject): List<String?> {
        val results = mutableListOf<String>()
        collectTextCandidates(major, textKeyPatterns, maxDepth = 4, results)
        return results.distinctBy { it.trim().lowercase() }
            .filter { trimmed ->
                val t = trimmed.trim()
                t.isNotEmpty() && !t.contains("http://") && !t.contains("https://")
            }
    }

    private fun extractURLs(
        element: JsonElement,
        keys: Set<String>,
        maxDepth: Int,
        predicate: (String) -> Boolean,
    ): List<String> {
        val results = mutableListOf<String>()
        collectURLs(element, keys, maxDepth, predicate, results)
        return results.distinct()
    }

    private fun collectURLs(
        element: JsonElement,
        keys: Set<String>,
        maxDepth: Int,
        predicate: (String) -> Boolean,
        results: MutableList<String>,
    ) {
        if (maxDepth < 0) return

        when (element) {
            is JsonObject -> {
                for ((key, value) in element) {
                    if (key.lowercase() in keys && value is kotlinx.serialization.json.JsonPrimitive) {
                        val raw = value.content
                        if (predicate(raw)) {
                            normalizedURL(raw)?.let { results.add(it) }
                        }
                    }
                    collectURLs(value, keys, maxDepth - 1, predicate, results)
                }
            }
            is JsonArray -> {
                for (item in element) {
                    collectURLs(item, keys, maxDepth - 1, predicate, results)
                }
            }
            else -> {}
        }
    }

    private fun collectTextCandidates(
        element: JsonElement,
        keys: Set<String>,
        maxDepth: Int,
        results: MutableList<String>,
    ) {
        if (maxDepth < 0) return

        when (element) {
            is JsonObject -> {
                for ((key, value) in element) {
                    if (key.lowercase() in keys && value is kotlinx.serialization.json.JsonPrimitive) {
                        results.add(value.content)
                    }
                    collectTextCandidates(value, keys, maxDepth - 1, results)
                }
            }
            is JsonArray -> {
                for (item in element) {
                    collectTextCandidates(item, keys, maxDepth - 1, results)
                }
            }
            else -> {}
        }
    }

    private fun isLikelyNavigationURL(rawValue: String): Boolean {
        val url = normalizedURL(rawValue) ?: return false
        if (!url.contains("bilibili.com")) return false
        return !isLikelyImageURL(url)
    }

    private fun isLikelyImageURL(rawValue: String): Boolean {
        val url = normalizedURL(rawValue) ?: return false
        val lower = url.lowercase()
        if ("hdslb.com" in lower) return true
        val imageExtensions = listOf(".jpg", ".jpeg", ".png", ".webp", ".avif", ".gif", ".bmp")
        if (imageExtensions.any { lower.endsWith(it) }) return true
        return "/bfs/" in lower
    }

    private fun fallbackTargetURL(id: String, kind: DynamicCard.Kind): String = when (kind) {
        DynamicCard.Kind.VIDEO -> "https://t.bilibili.com/$id"
        else -> "https://www.bilibili.com/opus/$id"
    }

    companion object {
        private val urlKeyPatterns = setOf(
            "jump_url", "url", "link", "target_url", "target", "schema",
        )
        private val imageKeyPatterns = setOf(
            "cover", "src", "url", "image", "img", "pic", "poster",
        )
        private val textKeyPatterns = setOf(
            "title", "text", "content", "desc", "summary", "copy_text", "name",
        )
    }
}

File: Sources/FocusCore/DynamicFeedService.swift
import Foundation

public struct DynamicFeedService: Sendable {
    public struct FeedPage: Equatable, Sendable {
        public let cards: [DynamicCard]
        public let nextOffset: String?

        public init(cards: [DynamicCard], nextOffset: String?) {
            self.cards = cards
            self.nextOffset = nextOffset
        }
    }

    public enum ServiceError: Error, Equatable, Sendable, LocalizedError {
        case loginRequired
        case invalidResponse
        case api(code: Int, message: String)

        public var errorDescription: String? {
            switch self {
            case .loginRequired:
                return "需要登录或登录已失效"
            case .invalidResponse:
                return "动态接口返回无效数据"
            case let .api(code, message):
                return "动态接口失败（\(code)）：\(message)"
            }
        }
    }

    public typealias RequestLoader = @Sendable (URLRequest) async throws -> (Data, URLResponse)

    private let cookieProvider: any CookieSnapshotProvider
    private let requestLoader: RequestLoader
    private let timezoneOffsetMinutes: Int

    public init(
        cookieProvider: any CookieSnapshotProvider,
        timezoneOffsetMinutes: Int = -(TimeZone.current.secondsFromGMT() / 60),
        requestLoader: @escaping RequestLoader = { request in
            try await URLSession.shared.data(for: request)
        }
    ) {
        self.cookieProvider = cookieProvider
        self.timezoneOffsetMinutes = timezoneOffsetMinutes
        self.requestLoader = requestLoader
    }

    public func fetchFollowingFeed() async throws -> [DynamicCard] {
        try await fetchFollowingFeedPage().cards
    }

    public func fetchFollowingFeedPage(offset: String? = nil) async throws -> FeedPage {
        let cookies = await cookieProvider.loadCookies()
        guard !cookies.isEmpty else {
            throw ServiceError.loginRequired
        }

        var request = URLRequest(url: Self.endpoint(timezoneOffsetMinutes: timezoneOffsetMinutes, offset: offset))
        request.httpMethod = "GET"
        request.setValue("https://t.bilibili.com/", forHTTPHeaderField: "Referer")
        request.setValue("application/json, text/plain, */*", forHTTPHeaderField: "Accept")
        request.setValue("Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1", forHTTPHeaderField: "User-Agent")
        request = Self.attach(cookies: cookies, to: request)

        let (data, response) = try await requestLoader(request)
        guard
            let httpResponse = response as? HTTPURLResponse,
            (200..<300).contains(httpResponse.statusCode)
        else {
            throw ServiceError.invalidResponse
        }

        return try Self.decodeFeedPage(from: data)
    }

    public static func decodeCards(from data: Data) throws -> [DynamicCard] {
        try decodeFeedPage(from: data).cards
    }

    public static func decodeFeedPage(from data: Data) throws -> FeedPage {
        guard let payload = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw ServiceError.invalidResponse
        }

        let code = payload.intValue(at: ["code"]) ?? -1
        let message = payload.stringValue(at: ["message"]) ?? "unknown"

        if code == -101 {
            throw ServiceError.loginRequired
        }

        guard code == 0 else {
            throw ServiceError.api(code: code, message: message)
        }

        guard let items = payload.arrayValue(at: ["data", "items"]) else {
            return FeedPage(cards: [], nextOffset: nil)
        }

        let cards: [DynamicCard] = items.compactMap { item -> DynamicCard? in
            guard let dictionary = item as? [String: Any] else {
                return nil
            }
            return makeCard(from: dictionary)
        }

        let nextOffset = payload.stringValue(at: ["data", "offset"])?.trimmingCharacters(in: .whitespacesAndNewlines)
        return FeedPage(cards: cards, nextOffset: nextOffset?.isEmpty == true ? nil : nextOffset)
    }

    public static func attach(cookies: [HTTPCookie], to request: URLRequest) -> URLRequest {
        guard !cookies.isEmpty else {
            return request
        }

        var updatedRequest = request
        let headerFields = HTTPCookie.requestHeaderFields(with: cookies)
        for (name, value) in headerFields {
            updatedRequest.setValue(value, forHTTPHeaderField: name)
        }
        return updatedRequest
    }

    private static func endpoint(timezoneOffsetMinutes: Int, offset: String?) -> URL {
        var components = URLComponents(string: "https://api.bilibili.com/x/polymer/web-dynamic/v1/feed/all")!
        var queryItems = [
            URLQueryItem(name: "type", value: "all"),
            URLQueryItem(name: "timezone_offset", value: String(timezoneOffsetMinutes)),
        ]
        if let offset, !offset.isEmpty {
            queryItems.append(URLQueryItem(name: "offset", value: offset))
        }
        components.queryItems = queryItems
        return components.url!
    }

    private static func makeCard(from item: [String: Any]) -> DynamicCard? {
        let modules = item.dictionaryValue(at: ["modules"]) ?? [:]
        let author = modules.dictionaryValue(at: ["module_author"]) ?? [:]
        let dynamic = modules.dictionaryValue(at: ["module_dynamic"]) ?? [:]
        let major = dynamic.dictionaryValue(at: ["major"]) ?? [:]
        let basic = item.dictionaryValue(at: ["basic"]) ?? [:]

        guard
            let id = item.stringValue(at: ["id_str"])
                ?? basic.stringValue(at: ["comment_id_str"])
                ?? basic.stringValue(at: ["rid_str"])
                ?? basic.intValue(at: ["comment_id"]).map(String.init)
        else {
            return nil
        }

        let authorName = author.stringValue(at: ["name"]) ?? "Bilibili"
        let authorMID = author.intValue(at: ["mid"]).map(Int64.init) ?? 0
        let authorAvatar = normalizedURL(author.stringValue(at: ["face"]))
        let publishTime = author.stringValue(at: ["pub_time"])
            ?? author.stringValue(at: ["pub_action"])
            ?? author.stringValue(at: ["pub_time_label"])
            ?? ""

        let majorType = major.stringValue(at: ["type"]) ?? ""
        let jumpCandidates = [
            basic.stringValue(at: ["jump_url"]),
            major.stringValue(at: ["archive", "jump_url"]),
            major.stringValue(at: ["article", "jump_url"]),
            major.stringValue(at: ["pgc", "jump_url"]),
            major.stringValue(at: ["courses", "jump_url"]),
            major.stringValue(at: ["music", "jump_url"]),
            major.stringValue(at: ["medialist", "jump_url"]),
            major.stringValue(at: ["live", "jump_url"]),
            major.stringValue(at: ["opus", "jump_url"]),
            major.stringValue(at: ["common", "jump_url"]),
        ]
        .compactMap(normalizedURL(_:))
        + extractJumpURLs(from: major)

        let coverURLs = makeCoverURLs(major: major)
        let text = makeText(dynamic: dynamic, major: major)
        let videoURL = jumpCandidates.first(where: isVideoLikeURL(_:))
        let kind = makeKind(
            majorType: majorType,
            itemType: item.stringValue(at: ["type"]) ?? "",
            videoURL: videoURL,
            coverURLs: coverURLs
        )
        let targetURL = videoURL
            ?? jumpCandidates.first
            ?? fallbackTargetURL(for: id, kind: kind)

        return DynamicCard(
            id: id,
            kind: kind,
            author: .init(mid: authorMID, name: authorName, avatarURL: authorAvatar),
            publishTime: publishTime,
            text: text,
            coverURLs: coverURLs,
            targetURL: targetURL,
            videoURL: videoURL
        )
    }

    private static func makeText(dynamic: [String: Any], major: [String: Any]) -> String {
        let candidates = [
            dynamic.stringValue(at: ["desc", "text"]),
            major.stringValue(at: ["opus", "summary", "text"]),
            major.stringValue(at: ["archive", "title"]),
            major.stringValue(at: ["article", "title"]),
            major.stringValue(at: ["pgc", "title"]),
            major.stringValue(at: ["courses", "title"]),
            major.stringValue(at: ["music", "title"]),
            major.stringValue(at: ["medialist", "title"]),
            major.stringValue(at: ["live", "title"]),
            major.stringValue(at: ["common", "title"]),
        ] + extractTextCandidates(from: major)

        return candidates
            .compactMap { $0?.trimmingCharacters(in: .whitespacesAndNewlines) }
            .first(where: { !$0.isEmpty }) ?? ""
    }

    private static func makeCoverURLs(major: [String: Any]) -> [URL] {
        var urls: [URL] = []

        if let archiveCover = normalizedURL(major.stringValue(at: ["archive", "cover"])) {
            urls.append(archiveCover)
        }

        if let articleCovers = major.arrayValue(at: ["article", "covers"]) {
            urls.append(contentsOf: articleCovers.compactMap { cover in
                guard let value = cover as? String else { return nil }
                return normalizedURL(value)
            })
        }

        if let pgcCover = normalizedURL(
            major.stringValue(at: ["pgc", "cover"])
                ?? major.stringValue(at: ["pgc", "ep_cover"])
        ) {
            urls.append(pgcCover)
        }

        if let courseCover = normalizedURL(major.stringValue(at: ["courses", "cover"])) {
            urls.append(courseCover)
        }

        if let musicCover = normalizedURL(major.stringValue(at: ["music", "cover"])) {
            urls.append(musicCover)
        }

        if let medialistCover = normalizedURL(major.stringValue(at: ["medialist", "cover"])) {
            urls.append(medialistCover)
        }

        if let liveCover = normalizedURL(
            major.stringValue(at: ["live", "cover"])
                ?? major.stringValue(at: ["live", "room_cover"])
        ) {
            urls.append(liveCover)
        }

        if let drawItems = major.arrayValue(at: ["draw", "items"]) {
            urls.append(contentsOf: drawItems.compactMap { item in
                guard let dictionary = item as? [String: Any] else { return nil }
                return normalizedURL(
                    dictionary.stringValue(at: ["src"])
                        ?? dictionary.stringValue(at: ["url"])
                )
            })
        }

        if let opusPictures = major.arrayValue(at: ["opus", "pics"]) {
            urls.append(contentsOf: opusPictures.compactMap { picture in
                guard let dictionary = picture as? [String: Any] else { return nil }
                return normalizedURL(
                    dictionary.stringValue(at: ["url"])
                        ?? dictionary.stringValue(at: ["src"])
                )
            })
        }

        if let commonCover = normalizedURL(major.stringValue(at: ["common", "cover"])) {
            urls.append(commonCover)
        }

        urls.append(contentsOf: extractImageURLs(from: major))

        var deduplicated: [URL] = []
        var seen = Set<String>()
        for url in urls {
            let key = url.absoluteString
            guard seen.insert(key).inserted else {
                continue
            }
            deduplicated.append(url)
        }
        return deduplicated
    }

    private static func makeKind(
        majorType: String,
        itemType: String,
        videoURL: URL?,
        coverURLs: [URL]
    ) -> DynamicCard.Kind {
        if videoURL != nil
            || itemType.contains("_AV")
            || [
                "MAJOR_TYPE_ARCHIVE",
                "MAJOR_TYPE_PGC",
                "MAJOR_TYPE_COURSES",
                "MAJOR_TYPE_MEDIALIST",
            ].contains(majorType)
        {
            return .video
        }

        switch majorType {
        case "MAJOR_TYPE_DRAW", "MAJOR_TYPE_OPUS":
            return coverURLs.isEmpty ? .text : .image
        case "MAJOR_TYPE_ARTICLE":
            return .articleLike
        default:
            return coverURLs.isEmpty ? .text : .image
        }
    }

    private static func normalizedURL(_ rawValue: String?) -> URL? {
        guard var rawValue else {
            return nil
        }

        rawValue = rawValue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !rawValue.isEmpty else {
            return nil
        }

        if rawValue.hasPrefix("//") {
            rawValue = "https:\(rawValue)"
        } else if rawValue.hasPrefix("/opus/")
            || rawValue.hasPrefix("/video/")
            || rawValue.hasPrefix("/bangumi/play/")
        {
            rawValue = "https://www.bilibili.com\(rawValue)"
        } else if rawValue.hasPrefix("/") {
            rawValue = "https://t.bilibili.com\(rawValue)"
        }

        guard let url = URL(string: rawValue) else {
            return nil
        }

        return FocusNavigationPolicy.canonicalWebURL(for: url)
    }

    private static func isVideoLikeURL(_ url: URL) -> Bool {
        let path = url.path.lowercased()
        return path.hasPrefix("/video/") || path.hasPrefix("/bangumi/play/")
    }

    private static func extractJumpURLs(from value: Any) -> [URL] {
        extractURLs(
            from: value,
            matchingKeys: ["jump_url", "url", "link", "target_url", "target", "schema"],
            maxDepth: 4,
            predicate: isLikelyNavigationURL(_:)
        )
    }

    private static func extractImageURLs(from value: Any) -> [URL] {
        extractURLs(
            from: value,
            matchingKeys: ["cover", "src", "url", "image", "img", "pic", "poster"],
            maxDepth: 4,
            predicate: isLikelyImageURL(_:)
        )
    }

    private static func extractTextCandidates(from value: Any) -> [String?] {
        var results: [String] = []
        collectTextCandidates(
            from: value,
            matchingKeys: ["title", "text", "content", "desc", "summary", "copy_text", "name"],
            maxDepth: 4,
            into: &results
        )

        var deduplicated: [String?] = []
        var seen = Set<String>()
        for text in results {
            let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
            guard
                !trimmed.isEmpty,
                !trimmed.contains("http://"),
                !trimmed.contains("https://"),
                seen.insert(trimmed).inserted
            else {
                continue
            }
            deduplicated.append(trimmed)
        }

        return deduplicated
    }

    private static func extractURLs(
        from value: Any,
        matchingKeys: Set<String>,
        maxDepth: Int,
        predicate: (String) -> Bool
    ) -> [URL] {
        var results: [URL] = []
        collectURLs(
            from: value,
            matchingKeys: matchingKeys,
            maxDepth: maxDepth,
            predicate: predicate,
            into: &results
        )

        var deduplicated: [URL] = []
        var seen = Set<String>()
        for url in results {
            guard seen.insert(url.absoluteString).inserted else {
                continue
            }
            deduplicated.append(url)
        }
        return deduplicated
    }

    private static func collectURLs(
        from value: Any,
        matchingKeys: Set<String>,
        maxDepth: Int,
        predicate: (String) -> Bool,
        into results: inout [URL]
    ) {
        guard maxDepth >= 0 else {
            return
        }

        if let dictionary = value as? [String: Any] {
            for (key, nestedValue) in dictionary {
                if matchingKeys.contains(key.lowercased()),
                   let rawValue = nestedValue as? String,
                   predicate(rawValue),
                   let url = normalizedURL(rawValue)
                {
                    results.append(url)
                }

                collectURLs(
                    from: nestedValue,
                    matchingKeys: matchingKeys,
                    maxDepth: maxDepth - 1,
                    predicate: predicate,
                    into: &results
                )
            }
            return
        }

        if let array = value as? [Any] {
            for nestedValue in array {
                collectURLs(
                    from: nestedValue,
                    matchingKeys: matchingKeys,
                    maxDepth: maxDepth - 1,
                    predicate: predicate,
                    into: &results
                )
            }
        }
    }

    private static func collectTextCandidates(
        from value: Any,
        matchingKeys: Set<String>,
        maxDepth: Int,
        into results: inout [String]
    ) {
        guard maxDepth >= 0 else {
            return
        }

        if let dictionary = value as? [String: Any] {
            for (key, nestedValue) in dictionary {
                if matchingKeys.contains(key.lowercased()),
                   let text = nestedValue as? String
                {
                    results.append(text)
                }

                collectTextCandidates(
                    from: nestedValue,
                    matchingKeys: matchingKeys,
                    maxDepth: maxDepth - 1,
                    into: &results
                )
            }
            return
        }

        if let array = value as? [Any] {
            for nestedValue in array {
                collectTextCandidates(
                    from: nestedValue,
                    matchingKeys: matchingKeys,
                    maxDepth: maxDepth - 1,
                    into: &results
                )
            }
        }
    }

    private static func isLikelyNavigationURL(_ rawValue: String) -> Bool {
        guard let url = normalizedURL(rawValue) else {
            return false
        }

        let host = url.host?.lowercased() ?? ""
        guard host.contains("bilibili.com") else {
            return false
        }

        return !isLikelyImageURL(url.absoluteString)
    }

    private static func isLikelyImageURL(_ rawValue: String) -> Bool {
        guard let url = normalizedURL(rawValue) else {
            return false
        }

        let host = url.host?.lowercased() ?? ""
        let path = url.path.lowercased()
        let imageExtensions = [".jpg", ".jpeg", ".png", ".webp", ".avif", ".gif", ".bmp"]

        if host.contains("hdslb.com") {
            return true
        }

        if imageExtensions.contains(where: path.hasSuffix(_:)) {
            return true
        }

        return path.contains("/bfs/")
    }

    private static func fallbackTargetURL(for id: String, kind: DynamicCard.Kind) -> URL {
        switch kind {
        case .video:
            return URL(string: "https://t.bilibili.com/\(id)")!
        case .text, .image, .articleLike:
            return URL(string: "https://www.bilibili.com/opus/\(id)")!
        }
    }
}

private extension Dictionary where Key == String, Value == Any {
    func value(at path: [String]) -> Any? {
        guard let key = path.first else {
            return nil
        }

        let nextValue = self[key]
        guard path.count > 1 else {
            return nextValue
        }

        guard let nestedDictionary = nextValue as? [String: Any] else {
            return nil
        }

        return nestedDictionary.value(at: Array(path.dropFirst()))
    }

    func dictionaryValue(at path: [String]) -> [String: Any]? {
        value(at: path) as? [String: Any]
    }

    func arrayValue(at path: [String]) -> [Any]? {
        value(at: path) as? [Any]
    }

    func stringValue(at path: [String]) -> String? {
        if let string = value(at: path) as? String {
            return string
        }

        if let number = value(at: path) as? NSNumber {
            return number.stringValue
        }

        return nil
    }

    func intValue(at path: [String]) -> Int? {
        if let int = value(at: path) as? Int {
            return int
        }

        if let number = value(at: path) as? NSNumber {
            return number.intValue
        }

        if let string = value(at: path) as? String {
            return Int(string)
        }

        return nil
    }
}
